### PR TITLE
Go sec pipeline components

### PR DESCRIPTION
**Pull request by andream16**
Closes SAAS-752

# What's new
- Base Dockerfile to be used to build components as a base.
- Gosec scanner: hopefully most of this code will be re-used to have a `sarif-to-ocsf` package
- Json Logger reporter
- Custom Annotation enricher

# Updates
- Bumped dep on sdk for the git-clone to leverage a proper tag
- Using own sarif package
- Leveraging proto contract to encode Data Sources as discussed

# Things that I'm not tackling here
- There's some funkyness with CGO=0 + SQLite. I think that we need to add asap support for remote storage so the containers can speak over the same storage. Right now I'm not building with CGO disabled to work around it for testing. Will come back to it.
- Will find a better way to build the binaries from this base image. I assume this will change as soon as I start build and pushing images so this is a WIP to move forward right now.

# Related
- https://github.com/smithy-security/smithy/pull/488
- https://github.com/smithy-security/pkg/pull/2

# What comes next
- Building | Packaging > Pushing > Distributing
- Better storage integration for "remote" setup - also useful for docker networks for when we write e2e tests

### CODE REVIEW

File: new-components/scanners/gosec/internal/sarif/sarif.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+const (
+	// GoSecOutFormatJson tells the parser to unmarshal the raw data using json umarshaler.
+	GoSecOutFormatJson GoSecOutFormat = "json"
+)
+
+type (
+	// GosecOut represents the sarif output returned by running https://github.com/securego/gosec.
+	GosecOut struct {
```
Comment (northdpole) [inline new-components/scanners/gosec/internal/sarif/sarif.go:23]: this is not sarif, i think gosec either has a bug or you didn't produce output in the correct format

File: new-components/scanners/gosec/internal/sarif/sarif.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+)
+
+func (g GoSecOutFormat) String() string {
+	return string(g)
+}
+
+// NewParserFromEnv returns a new Parsed by doing environment variables lookup.
+func NewParserFromEnv() (*Parser, error) {
+	parseFormat, err := env.GetOrDefault(
+		"GOSEC_OUT_FORMAT",
```
Comment (northdpole) [inline new-components/scanners/gosec/internal/sarif/sarif.go:66]: we should only support a single output format per tool, if it supports something interoperable then that's the preference otherwise we deal with their json but please, supporting absolutely everything is boiling the ocean

File: new-components/enrichers/custom-annotation/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+The component used environment variables for configuration.
+
+It requires the component
+environment variables defined [here](https://github.com/smithy-security/smithy/blob/main/sdk/README.md#component) as well as the following:
+
+| Environment Variable       | Type   | Required | Default | Description                                 |
+|----------------------------|--------|----------|---------|---------------------------------------------|
+| CUSTOM\_ANNOTATION\_NAME     | string | yes      | -       | The name of the annotation to be added.     |
+| CUSTOM\_ANNOTATION\_VALUES   | string | yes      | {}      | Json annotations to be added as annotation. |
```
Comment (dlicheva) [inline new-components/enrichers/custom-annotation/README.md:16]: It might be clearer to say they are in a key-value format and say e.g. `{foo:bar}`

File: new-components/scanners/gosec/internal/sarif/sarif.go (hunk 1 above)
Comment (andream16) [reply new-components/scanners/gosec/internal/sarif/sarif.go]: Refactored.

File: new-components/enrichers/custom-annotation/README.md (hunk 3 above)
Comment (andream16) [reply new-components/enrichers/custom-annotation/README.md]: Done

File: new-components/reporters/json-logger/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# json-logger
+
+This is component implements a [reporter](https://github.com/smithy-security/smithy/blob/main/sdk/component/component.go)
```
Comment (northdpole) [inline new-components/reporters/json-logger/README.md:3]: nitpick: typo

File: new-components/reporters/json-logger/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+that prints vulnerability findings as json.
+
+## Environment variables
+
+The component used environment variables for configuration.
```
Comment (northdpole) [inline new-components/reporters/json-logger/README.md:8]: nitpick: typo

File: new-components/enrichers/custom-annotation/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
+# custom-annotation
+
+This is component implements an [enricher](https://github.com/smithy-security/smithy/blob/main/sdk/component/component.go)
```
Comment (northdpole) [inline new-components/enrichers/custom-annotation/README.md:3]: nitpick: typo

File: new-components/enrichers/custom-annotation/.env [hunk 7]
```diff
@@ -0,0 +1,5 @@
+SMITHY_INSTANCE_ID=8d719c1c-c569-4078-87b3-4951bd4012ee
+SMITHY_LOG_LEVEL=debug
+SMITHY_BACKEND_STORE_TYPE=local
+CUSTOM_ANNOTATION_NAME=reachability
+CUSTOM_ANNOTATION_VALUES={"foo":"bar"}
```
Comment (northdpole) [inline new-components/enrichers/custom-annotation/.env:5]: this is used only for tests, correct? does it make sense to be in a dedicated dir or have a relevant comment 
so it's clear how the example components work?

File: new-components/scanners/gosec/internal/transformer/parser.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+// Parse parses the raw gosec data into a known type.
+func (p *parser) Parse() (*sarif.Report, error) {
+	b, err := os.ReadFile(p.rawOutFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file '%s': %w", p.rawOutFilePath, err)
+	}
+
+	// TODO - BEGIN
+	// This will be replaced by https://linear.app/smithy/issue/OCU-277/support-openapi-spec-for-sarif.
```
Comment (northdpole) [inline new-components/scanners/gosec/internal/transformer/parser.go:39]: nit: wanna replace this with the equivalent github issue for visibility? this is a public repo and not eveyone has access to linear

**Review by northdpole — APPROVED** [review]

File: new-components/enrichers/custom-annotation/.env (hunk 7 above)
Comment (andream16) [reply new-components/enrichers/custom-annotation/.env:6]: Done with comments

File: new-components/enrichers/custom-annotation/README.md (hunk 6 above)
Comment (andream16) [reply new-components/enrichers/custom-annotation/README.md]: Fixed

File: new-components/reporters/json-logger/README.md (hunk 5 above)
Comment (andream16) [reply new-components/reporters/json-logger/README.md]: Fixed.

File: new-components/reporters/json-logger/README.md (hunk 4 above)
Comment (andream16) [reply new-components/reporters/json-logger/README.md]: Fixed.

**Review by northdpole — APPROVED** [review]

File: new-components/enrichers/custom-annotation/go.mod
Comment (ptzianos) [inline new-components/enrichers/custom-annotation/go.mod:1]: can you please move these changes to a separate commit for more clarity?

File: new-components/enrichers/custom-annotation/internal/annotation/annotation.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	annotationValues, err := env.GetOrDefault("CUSTOM_ANNOTATION_VALUES", "{}")
+	if err != nil {
+		return nil, err
+	}
+
+	if len(annotationValues) > 0 && annotationValues != "{}" {
```
Comment (ptzianos) [inline new-components/enrichers/custom-annotation/internal/annotation/annotation.go:52]: I think we can simplify this a bit more and just unmarshall if the annotationValues is a non-empty string

File: new-components/enrichers/custom-annotation/internal/annotation/annotation.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// Annotate adds annotated values to passed findings.
+func (ca *customAnnotator) Annotate(
+	ctx context.Context,
+	findings []*ocsf.VulnerabilityFinding,
+) ([]*ocsf.VulnerabilityFinding, error) {
+	var (
+		providerName = "custom-annotation-enricher"
+		newFindings  = slices.Clone(findings)
```
Comment (ptzianos) [inline new-components/enrichers/custom-annotation/internal/annotation/annotation.go:81]: why do this cloning over here?

File: new-components/enrichers/custom-annotation/internal/annotation/annotation.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+				With(slog.Int("num_findings", len(findings))).
+				With(slog.String("provider_name", providerName)).
+				With(slog.String("annotation_name", ca.conf.AnnotationName)).
+				With(slog.Any("annotation_values", ca.conf.AnnotationValues))
+	)
+
+	logger.Debug("preparing to annotate findings...")
+
+	for idx := range newFindings {
+		b, err := json.Marshal(ca.conf.AnnotationValues)
```
Comment (ptzianos) [inline new-components/enrichers/custom-annotation/internal/annotation/annotation.go:92]: this unmarshaling should be done once in the NewConf instead of being repeated for every loop since the resulting values will be the same

File: new-components/enrichers/custom-annotation/internal/annotation/annotation_test.go [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+	annotationValuesAsJson, err := json.Marshal(annotationValues)
+	require.NoError(t, err)
+
+	defer cancel()
+
+	t.Run("it should enrich the issues successfully", func(t *testing.T) {
+		annotator, err := annotation.NewCustomAnnotator(conf)
+		require.NoError(t, err)
+
+		enrichedVulns, err := annotator.Annotate(ctx, vulns)
```
Comment (ptzianos) [inline new-components/enrichers/custom-annotation/internal/annotation/annotation_test.go:226]: I think you can use the `.ElementsMatch` function to check the lists more easily over here

File: new-components/reporters/json-logger/cmd/main_test.go [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+func (c *capturingLogger) Error(msg string, keyvals ...any) {
+	c.capture(msg, keyvals...)
+}
+
+func (c *capturingLogger) With(args ...any) component.Logger {
+	c.capture("", args...)
+	return c
+}
+
+func ptr[T any](v T) *T {
```
Comment (ptzianos) [inline new-components/reporters/json-logger/cmd/main_test.go:96]: you can use ptr.Ptr instead of this

File: new-components/enrichers/custom-annotation/go.mod
Comment (andream16) [reply new-components/enrichers/custom-annotation/go.mod:1]: Done for all new vendors

File: new-components/scanners/gosec/go.mod
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/go.mod:1]: Done for all vendors / modules

File: new-components/scanners/gosec/README.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+The component uses environment variables for configuration.
+
+It requires the component
+environment variables defined [here](https://github.com/smithy-security/smithy/blob/main/sdk/README.md#component) as well
+as the following:
+
+| Environment Variable     | Type   | Required | Default    | Description                                             |
+|--------------------------|--------|----------|------------|---------------------------------------------------------|
+| GOSEC\_RAW\_OUT\_FILE\_PATH  | string | yes      | -          | The path where to find the gosec report                 |
+| GOSEC\_TARGET\_TYPE         | string | false    | repository | The type of target that was used to generate the report |
```
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/README.md:18]: I was thinking that gosec can also get in input a ready report - based on the curl target - that's why I kept it customisable

File: new-components/scanners/gosec/internal/transformer/transformer_test.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+		transformer.TargetTypeRepository,
+		transformer.GosecTransformerWithClock(clock),
+	)
+	require.NoError(t, err)
+
+	t.Run("it should transform correctly the finding to ocsf format", func(t *testing.T) {
+		findings, err := ocsfTransformer.Transform(ctx)
+		require.NoError(t, err)
+		require.NotEmpty(t, findings)
+		assert.Len(t, findings, 21)
```
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/internal/transformer/transformer_test.go:45]: The assertions below will be fine even with a different number of findings but I think it makes sense to halt it here - I can use a require 👌🏻

File: new-components/enrichers/custom-annotation/internal/annotation/annotation.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+
+	return cfg, nil
+}
+
+// NewCustomAnnotator returns a new custom annotator.
+func NewCustomAnnotator(conf *Conf) (*customAnnotator, error) {
+	switch {
+	case conf == nil:
+		return nil, errors.New("nil annotation configuration")
+	case conf.AnnotationName == "":
```
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/enrichers/custom-annotation/internal/annotation/annotation.go:66]: I see your point - then we can just let the enricher exit early if this is the case

File: new-components/enrichers/custom-annotation/internal/annotation/annotation.go (hunk 10 above)
Comment (andream16) [reply new-components/enrichers/custom-annotation/internal/annotation/annotation.go:81]: I wanted to log the old vs new findings but didn't end up doing it - good spot - I'll remove it

File: new-components/enrichers/custom-annotation/internal/annotation/annotation_test.go [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	ocsf "github.com/smithy-security/smithy/sdk/gen/com/github/ocsf/ocsf_schema/v1"
+
+	"github.com/smithy-security/smithy/new-components/enrichers/custom-annotation/internal/annotation"
+)
+
+func ptr[T any](v T) *T {
```
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/enrichers/custom-annotation/internal/annotation/annotation_test.go:19]: I created that function for gosec only since it manipulates the actual `ocsf.VulnerabilityFinding` extensively using pointers (sad face here btw, very bad to use) - here I only needed it for tests

File: new-components/reporters/json-logger/cmd/main_test.go (hunk 13 above)
Comment (andream16) [reply new-components/reporters/json-logger/cmd/main_test.go:96]: As mentioned in another comment, here we only need this in the test

File: new-components/scanners/gosec/internal/transformer/transformer.go [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+
+	return "", errors.New("missing location information in finding")
+}
+
+func (g *gosecTransformer) mapConfidence(
+	ruleID string,
+	ruleToTools map[string]schemav1.ReportingDescriptor,
+) ocsf.VulnerabilityFinding_ConfidenceId {
+	toolRule, ok := ruleToTools[ruleID]
+	if ok && toolRule.Properties != nil {
```
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/internal/transformer/transformer.go:296]: I will simplify this

File: new-components/scanners/gosec/internal/transformer/transformer.go [hunk 19]
```diff
@@ -0,0 +1,391 @@
...
+	return ocsf.VulnerabilityFinding_CONFIDENCE_ID_UNKNOWN
+}
+
+func (g *gosecTransformer) mapTitleDesc(
+	res schemav1.Result,
+	ruleToTools map[string]schemav1.ReportingDescriptor,
+) (string, string) {
+	title, descr := *res.Message.Text, *res.Message.Text
+
+	rule, ok := ruleToTools[*res.RuleId]
```
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/internal/transformer/transformer.go:335]: Done

File: new-components/enrichers/custom-annotation/internal/annotation/annotation.go (hunk 10 above)
Comment (andream16) [reply new-components/enrichers/custom-annotation/internal/annotation/annotation.go]: Done

File: new-components/scanners/gosec/internal/transformer/transformer.go [hunk 20]
```diff
@@ -0,0 +1,391 @@
...
+	"errors"
+	"fmt"
+	"log/slog"
+	"os"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/smithy-security/pkg/env"
+	"google.golang.org/protobuf/encoding/protojson"
+
+	schemav1 "github.com/smithy-security/pkg/sarif/spec/gen/sarif-schema/v2-1-0"
```
[the remark this answers is not in the document]
Comment (ptzianos) [reply new-components/scanners/gosec/internal/transformer/transformer.go:14]: sorry I meant if we can use `sarifschemav210` for the import alias, because now we are calling it `schemav1` but its `v2.1.0` that we are importing

File: new-components/scanners/gosec/internal/transformer/transformer.go [hunk 21]
```diff
@@ -0,0 +1,391 @@
...
+)
+
+func (tt TargetType) String() string {
+	return string(tt)
+}
+
+// GosecTransformerWithClock allows customising the underlying clock.
+func GosecTransformerWithClock(clock clockwork.Clock) GosecTransformerOption {
+	return func(g *gosecTransformer) error {
+		if clock == nil {
```
[the remark this answers is not in the document]
Comment (ptzianos) [reply new-components/scanners/gosec/internal/transformer/transformer.go:46]: what is have in mind is this:
```golang

var ErrInvalidClock = errors.New("invalid nil clock")

func GosecTransformerWithClock(clock clockwork.Clock) (GosecTransformerOption, error) {
	if clock == nil {
		return nil, ErrInvalidClock
	}
	return func(g *gosecTransformer) error {
		g.clock = clock
		return nil
	}
}
```
This way if the clock is nil, the function will fail fast instead of failing when it's invoked inside the transformer constructor

File: new-components/scanners/gosec/internal/transformer/transformer.go [hunk 22]
```diff
@@ -0,0 +1,391 @@
...
+			return fmt.Errorf("invalid nil clock")
```
[the remark this answers is not in the document]
Comment (ptzianos) [reply new-components/scanners/gosec/internal/transformer/transformer.go:47]: I prefer if we use `go-errors` everywhere for consistency but also accurately capturing the error stack traces from wherever they are coming from

File: new-components/scanners/gosec/internal/transformer/transformer.go [hunk 23]
```diff
@@ -0,0 +1,391 @@
...
+		}
+		g.clock = clock
+		return nil
+	}
+}
+
+// NewTransformerFromEnv returns a new transformer initialised from environment variables lookup.
+func NewTransformerFromEnv(opts ...GosecTransformerOption) (*gosecTransformer, error) {
```
[the remark this answers is not in the document]
Comment (ptzianos) [reply new-components/scanners/gosec/internal/transformer/transformer.go:55]: not too strong of an opinion tbh, but I would prefer if we kept a consistent style, meaning that we can have some sane defaults and then allow the user to modify them by supplying options. for the tests we can have a private `newTransformer` function that returns a sanely configured object that can be used by tests that is then modified by the `GosecTransformerOption`s passed to the publicly available `New`

File: new-components/enrichers/custom-annotation/internal/annotation/annotation_test.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (ptzianos) [reply new-components/enrichers/custom-annotation/internal/annotation/annotation_test.go:18]: how about we add this to `github.com/smithy-security/smithy/utils`?

File: new-components/scanners/gosec/internal/transformer/transformer.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/internal/transformer/transformer.go]: Simplified.

File: new-components/scanners/gosec/internal/transformer/transformer.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/internal/transformer/transformer.go:45]: I think that calling this before the designed loop for function would add more code for little gains

This pattern works quite well, is well known and follows the standard that I made in the changes

Some sources:
- https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
- https://michalzalecki.com/golang-options-pattern/

If you don't mind I'll keep this as is :)

File: new-components/scanners/gosec/internal/transformer/transformer.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (andream16) [reply new-components/scanners/gosec/internal/transformer/transformer.go]: Sure, done

**Review by ptzianos — APPROVED** [review]
